patch_subset/py/patch_subset_method.py

```diff
@@ -0,0 +1,183 @@
+asking a server for a set of codepoints. The server responds with a patch that
+the client applies to their copy of the font.
+
+This is a python wrapper around a C++ implementation. Uses python ctypes to
+interface with the C++ code.
+"""
+import collections
+from ctypes import byref
+from ctypes import c_bool
+from ctypes import c_char_p
+from ctypes import c_uint32
+from ctypes import c_ubyte
+from ctypes import c_void_p
+from ctypes import cdll
+from ctypes import POINTER
+from ctypes import Structure
+
+from analysis import request_graph
+
+# Define the C interface
+patch_subset = cdll.LoadLibrary('./patch_subset/py/patch_subset_session.so')  # pylint: disable=invalid-name
+
+
+class RECORD(Structure):
+  _fields_ = [("request_size", c_uint32), ("response_size", c_uint32)]
+
+
+new_session = patch_subset.PatchSubsetSession_new  # pylint: disable=invalid-name
+new_session.restype = c_void_p
+extend = patch_subset.PatchSubsetSession_extend  # pylint: disable=invalid-name
+extend.restype = c_bool
+get_font_bytes = patch_subset.PatchSubsetSession_get_font  # pylint: disable=invalid-name
+get_font_bytes.restype = POINTER(c_ubyte)
+get_requests = patch_subset.PatchSubsetSession_get_requests  # pylint: disable=invalid-name
+get_requests.restype = POINTER(RECORD)
+
+Record = collections.namedtuple("Record", ["request_size", "response_size"])
+
+
+# Module methods
+def name():
+  return "PatchSubset_PFE"
+
+
+def start_session(font_directory):
+  """Starts a new PFE session for this method."""
+  return PatchSubsetPfeSession(font_directory)
+
+
+class PatchSubsetError(Exception):
+  """The patch subset c code failed to execute."""
+
+
+class FontSession:
+  """Tracks the PFE state for a single font.
+
+  Most importantly it keeps a record of what requests where sent
+  for each page view for a particular font.
+  """
+
+  def __init__(self, font_directory, font_id, page_view_count):
+    # TODO(garretrieger): set font directory correctly.
+    font_directory_c = c_char_p(font_directory.encode("utf-8"))
+    font_id_c = c_char_p(font_id.encode("utf-8"))
+    self.session = c_void_p(new_session(font_directory_c, font_id_c))
+    self.delete_session = patch_subset.PatchSubsetSession_delete
+    self.records_by_view = [[]] * page_view_count
+    self.last_record_index = None
+    self.page_view_count = page_view_count
+
+  def __del__(self):
+    if self.session:
+      self.delete_session(self.session)
+      self.session = None
+
+  def get_records_by_page_view(self, index):
+    return self.records_by_view[index]
+
+  def page_viewed(self):
+    self.page_view_count += 1
+    self.records_by_view.append([])
+
+  def extend(self, codepoints):
+    """Extends the tracked font to cover new codepoints.
+
+    Records any resulting requests needed to make the extension.
+    """
+    codepoint_array_c = (c_uint32 * len(codepoints))()
+    i = 0
+    for codepoint in codepoints:
```

Comment:
`for idx, codepoint in enumerate(codepoints): codepoint_array[idx] = codepoint` ?

Reply:
Fixed in upcoming PR.